inc/ptw.h

```diff
@@ -0,0 +1,142 @@
+#ifndef PTW_H
+#define PTW_H
+
+#include "cache.h"
+#include <map>
+
+#define IS_PSCL5 11
+#define IS_PSCL4 12
+#define IS_PSCL3 13
+#define IS_PSCL2 14
+
+#define PSCL5_SET 1
+#define PSCL5_WAY 2
+
+#define PSCL4_SET 1
+#define PSCL4_WAY 4
+
+#define PSCL3_SET 2
+#define PSCL3_WAY 4
+
+#define PSCL2_SET 4
+#define PSCL2_WAY 8
+
+#define PTW_RQ_SIZE STLB_MSHR_SIZE
+#define PTW_WQ_SIZE 0
+#define PTW_MSHR_SIZE 1
+#define PTW_PQ_SIZE 0
+
+#define MMU_CACHE_LATENCY 1
+
+#define NUM_ENTRIES_PER_PAGE 512
+
+#define IS_PTL1 1
+#define IS_PTL2 2
+#define IS_PTL3 3
+#define IS_PTL4 4
+#define IS_PTL5 5
+
+//Virtual Address: 57 bit (9+9+9+9+9+12), rest MSB bits will be used to generate a unique VA per CPU.
+//PTL5->PTL4->PTL3->PTL2->PTL1->PFN
+
+class PageTablePage
+{
+	public:
+		PageTablePage* entry[NUM_ENTRIES_PER_PAGE];
+		uint64_t next_level_base_addr[NUM_ENTRIES_PER_PAGE];
+
+	PageTablePage()
+	{
+		for(int i = 0; i < NUM_ENTRIES_PER_PAGE; i++)
+		{
+			entry[i] = NULL;
+			next_level_base_addr[i] = UINT64_MAX;
+		}
+	}
+
+	~PageTablePage()
+	{
+		for(int i = 0; i < NUM_ENTRIES_PER_PAGE; i++)
+		{
+			if(entry[i] != NULL)
+				delete(entry[i]);
+		}
+	}
+};
+
+class PageTableWalker : public MemoryRequestConsumer, public MemoryRequestProducer
+{
+	public:
+		const string NAME;
+		uint32_t cpu;
+		uint8_t cache_type;
+		uint32_t psc_latency;
+
+		uint8_t LATENCY;
+	
+		uint64_t rq_full;
+		uint64_t total_miss_latency;
+
+		uint64_t next_translation_virtual_address = 0xf000000f00000000;
+
+		std::map<uint64_t, uint64_t> page_table;
+
+		PACKET_QUEUE WQ{NAME+"_WQ", PTW_WQ_SIZE},
+					RQ{NAME+"_RQ", PTW_RQ_SIZE},
+					MSHR{NAME+"_MSHR", PTW_MSHR_SIZE},
+					PQ{NAME+"_PQ", PTW_PQ_SIZE};
+
+	CACHE PSCL5{"PSCL5", PSCL5_SET, PSCL5_WAY, 0, 0, 0, 0, 0, 0}, //Translation from L5->L4
+          PSCL4{"PSCL4", PSCL4_SET, PSCL4_WAY, 0, 0, 0, 0, 0, 0}, //Translation from L5->L3
+          PSCL3{"PSCL3", PSCL3_SET, PSCL3_WAY, 0, 0, 0, 0, 0, 0}, //Translation from L5->L2
+          PSCL2{"PSCL2", PSCL2_SET, PSCL2_WAY, 0, 0, 0, 0, 0, 0}; //Translation from L5->L1
```

Comment:
I am reusing CACHE class as I am also tracking stats for PSCs (Reset/Print/Record stats function). If someone want to add replacement/prefetcher for PSCs, then it can also be done with the CACHE class.

If required for this PR, I will create a new class for PSC and add these three functions.

Reply:
Ah, now I see what you were going for here. I still think the class is too heavyweight for your purposes. You're instantiating 4 objects of type `CACHE` that never reference the bulk of the code in cache.cc. I think you should go with something simpler. If it were me, I would make a `std::vector<BLOCK>` and track the stats locally. If you want to make a simple class for it, that works just as well.

If you're concerned about being able to use the `print_roi_stats()` family of functions, I would make a `PageTableWalker::print_stats()` that you can call at the end of the simulation.